rclrs/src/clock.rs

```diff
@@ -0,0 +1,170 @@
+use crate::rcl_bindings::*;
+use crate::{error::ToResult, time::Time, to_rclrs_result, RclrsError};
+use std::sync::{Arc, Mutex};
+
+/// Enum to describe clock type. Redefined for readability and to eliminate the uninitialized case
+/// from the `rcl_clock_type_t` enum in the binding.
+#[derive(Clone, Debug, Copy)]
+pub enum ClockType {
+    /// Time with behavior dependent on the `set_ros_time(bool)` function. If called with `true`
+    /// it will be driven by a manual value override, otherwise it will be System Time
+    RosTime = 1,
+    /// Wall time depending on the current system
+    SystemTime = 2,
+    /// Steady time, monotonically increasing but not necessarily equal to wall time.
+    SteadyTime = 3,
+}
+
+impl From<ClockType> for rcl_clock_type_t {
+    fn from(clock_type: ClockType) -> Self {
+        match clock_type {
+            ClockType::RosTime => rcl_clock_type_t::RCL_ROS_TIME,
+            ClockType::SystemTime => rcl_clock_type_t::RCL_SYSTEM_TIME,
+            ClockType::SteadyTime => rcl_clock_type_t::RCL_STEADY_TIME,
+        }
+    }
+}
+
+/// Struct that implements a Clock and wraps `rcl_clock_t`.
+pub struct Clock {
+    _type: ClockType,
+    _rcl_clock: Arc<Mutex<rcl_clock_t>>,
```

Comment:
I don't believe we ever need to change the value of the inner `rcl_clock_t`. I would recommend removing the `Mutex<>` wrapper and see if this code still compiles. If it does then we don't need the `Mutex<>`, and we'll avoid a lot of unnecessary locking.

Reply:
I looked at it and I couldn't come up with a way to remove the `Mutex` since the value in `_rcl_clock` could be read and written from different threads at the same time. Still, the `Arc` was unnecessary since an `rcl_clock_t` instance is always owned by a single clock so at least I could remove that layer of indirection https://github.com/ros2-rust/ros2_rust/pull/325/commits/19a73c3eb0a33e562d55f43024f3120964f91fbd.

Specifically, the `set_ros_time` [function](https://github.com/luca-della-vedova/ros2_rust/blob/19a73c3eb0a33e562d55f43024f3120964f91fbd/rclrs/src/clock.rs#L59-L74) would modify the inner `_rcl_clock` by setting its internal time, at the same time the `now` [function](https://github.com/luca-della-vedova/ros2_rust/blob/19a73c3eb0a33e562d55f43024f3120964f91fbd/rclrs/src/clock.rs#L77-L88) would read the internal clock and return it. I'm a bit afraid of what would happen if they were both called from concurrent threads, `rcl` documentation specifies that even though the functions use atomics, concurrent calls on the same `clock` object [are not safe](https://github.com/ros2/rcl/blob/960bb4ad353a43abfde5ecb831bd87823454bfc0/rcl/include/rcl/time.h#L645).